lib/writer.ml

```diff
@@ -154,58 +181,152 @@ let assemble_alpn_protocol p =
 let assemble_alpn_protocols protocols =
   assemble_list Two assemble_alpn_protocol protocols
 
+let assemble_supported_versions vs =
+  assemble_list One assemble_any_protocol_version vs
+
 let assemble_extension = function
   | `ECPointFormats formats ->
      (assemble_ec_point_formats formats, EC_POINT_FORMATS)
   | `SecureRenegotiation x ->
      let buf = create 1 in
      set_uint8 buf 0 (len x);
      (buf <+> x, RENEGOTIATION_INFO)
-  | `ExtendedMasterSecret -> (Cstruct.create 0, EXTENDED_MASTER_SECRET)
+  | `ExtendedMasterSecret -> (create 0, EXTENDED_MASTER_SECRET)
+  | _ -> invalid_arg "unknown extension"
+
+let assemble_cookie c =
+  let l = create 2 in
+  BE.set_uint16 l 0 (len c) ;
+  l <+> c
+
+let assemble_psk_key_exchange_mode mode =
+  let c = create 1 in
+  set_uint8 c 0 (psk_key_exchange_mode_to_int mode) ;
+  c
+
+let assemble_psk_key_exchange_modes modes =
+  assemble_list One assemble_psk_key_exchange_mode modes
+
+let assemble_ext (pay, typ) =
+  let buf = Cstruct.create 4 in
+  BE.set_uint16 buf 0 (extension_type_to_int typ);
+  BE.set_uint16 buf 2 (len pay);
+  buf <+> pay
+
+let assemble_extensions ?none_if_empty assemble_e es =
+  assemble_list ?none_if_empty Two assemble_e es
+
+let assemble_ca ca =
+  let lenbuf = create 2 in
+  let data = X509.Distinguished_name.encode_der ca in
+  BE.set_uint16 lenbuf 0 (len data) ;
+  lenbuf <+> data
+
+let assemble_certificate_authorities cas =
+  assemble_list Two assemble_ca cas
+
+let assemble_certificate_request_extension e =
+  assemble_ext @@ match e with
+  | `SignatureAlgorithms s ->
+    (assemble_signature_algorithms s, SIGNATURE_ALGORITHMS)
+  | `CertificateAuthorities cas ->
+    (assemble_certificate_authorities cas, CERTIFICATE_AUTHORITIES)
   | _ -> invalid_arg "unknown extension"
 
+let assemble_certificate_request_1_3 ?(context = Cstruct.empty) exts =
+  let clen = create 1 in
+  set_uint8 clen 0 (len context) ;
+  let exts = assemble_extensions assemble_certificate_request_extension exts in
+  clen <+> context <+> exts
+
 let assemble_client_extension e =
-  let pay, typ = match e with
-    | `EllipticCurves cs -> (assemble_elliptic_curves cs, ELLIPTIC_CURVES)
+  assemble_ext @@ match e with
+    | `SupportedGroups groups ->
+      (assemble_supported_groups groups, SUPPORTED_GROUPS)
     | `Hostname name -> (assemble_hostnames [name], SERVER_NAME)
     | `Padding x ->
-       let buf = create x in
-       memset buf 0 ;
-       (buf, PADDING)
+      let buf = create x in
+      memset buf 0 ;
```

Comment:
`Cstruct.create` already fills with null-bytes, as opposed to `Cstruct.create_unsafe`. Maybe this changed upstream since this was written? In any case, the `memset` here is redundant with newer versions of `Cstruct` (perhaps we could just constrain `tls` to require `>=` such version)?

Conversely we also have a lot of `Cstruct.create 1 ; Cstruct.set_uint8 b 0 ...` that could maybe be optimized a bit; not sure if the scary look of `Cstruct.create_unsafe` is worth it, and I suspect the allocations themselves are much more expensive than the redundant memsetting at the moment.